apps/backend/src/app/modules/users/users.service.spec.ts

```diff
@@ -0,0 +1,286 @@
+
+  beforeAll(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [UsersService, PRISMA_MOCK_PROVIDER]
+    })
+      .useMocker(mockDeep)
+      .compile();
+    usersService = module.get(UsersService);
+    db = module.get(EXTENDED_PRISMA_SERVICE);
+  });
+
+  it('should be defined', () => {
+    expect(usersService).toBeDefined();
+    expect(db).toBeDefined();
+  });
+
+  describe('getAll', () => {
+    it('should throw an error when passing both steamID and steamIDs', async () => {
+      const a: UsersGetAllQueryDto = {
+        steamID: '123456789',
+        steamIDs: ['123456789', '999999999']
+      };
+      await expect(usersService.getAll(a)).rejects.toThrow(BadRequestException);
+    });
+  });
+
+  describe('findOrCreateFromGame', () => {
+    it("should throw an error when the steamid's don't match", async () => {
+      const steamUserSummaryData: SteamUserSummaryData = {
+        avatar: '',
+        avatarfull: '',
+        avatarhash: '',
+        avatarmedium: '',
+        communityvisibilitystate: 0,
+        lastlogoff: 0,
+        loccountrycode: '',
+        personaname: '',
+        personastate: 0,
+        personastateflags: 0,
+        primaryclanid: '',
+        profilestate: 0,
+        profileurl: '',
+        realname: '',
+        timecreated: 0,
+        steamid: '123456789'
+      };
+      jest
+        .spyOn(usersService['steamService'], 'getSteamUserSummaryData')
+        .mockImplementationOnce(() => Promise.resolve(steamUserSummaryData));
+      await expect(
+        usersService.findOrCreateFromGame(BigInt(999999999))
+      ).rejects.toThrow(BadRequestException);
+    });
+    it('should throw an error when getSteamUserSummaryData throws and error', async () => {
+      jest
+        .spyOn(usersService['steamService'], 'getSteamUserSummaryData')
+        .mockImplementationOnce(() => Promise.reject('failed to start'));
+      await expect(
+        usersService.findOrCreateFromGame(BigInt(123456789))
+      ).rejects.toThrow(ServiceUnavailableException);
+    });
+    it('should call findOrCreateUser', async () => {
+      const steamUserSummaryData: SteamUserSummaryData = {
+        avatar: '',
+        avatarfull: '',
+        avatarhash: '',
+        avatarmedium: '',
+        communityvisibilitystate: 0,
+        lastlogoff: 0,
+        loccountrycode: '',
+        personaname: '',
+        personastate: 0,
+        personastateflags: 0,
+        primaryclanid: '',
+        profilestate: 0,
+        profileurl: '',
+        realname: '',
+        timecreated: 0,
+        steamid: '123456789'
+      };
+      jest
+        .spyOn(usersService['steamService'], 'getSteamUserSummaryData')
+        .mockImplementationOnce(() => Promise.resolve(steamUserSummaryData));
+      const spy = jest.spyOn(usersService, 'findOrCreateUser');
+      await usersService.findOrCreateFromGame(BigInt(123456789));
+      expect(spy).toHaveBeenCalled();
```

Comment:
Here and in some other instances below you can use `toHaveBeenCalledWith` and pass in a specific object, should check it matches. One of the major parts of these functions is creating the appropriate Prisma object so worth testing that.

Reply:
That's a good point. I originally decided against using `toHaveBeenCalledWith` because the method calls are type checked, but the type checking won't ensure we are using the right values within the type. I will go ahead and add those.

Reply:
Yeah, also remember TypeScript types are just compile-time, and we can even just disable them when getting in the way within tests by casting to `any`. `toHaveBeenCalledWith` will be just check that the underlying JavaScript objects match up, which will use strict type checks but *JavaScript* strict type checks so like, making sure a string is a string, number is a number, nothing more complex than that.